buildSrc/src/main/kotlin/Java9Modularity.kt

```diff
@@ -42,70 +49,146 @@ object Java9Modularity {
 
                 // derive the names of the source set and compile module task
                 val sourceSetName = defaultSourceSet.name + "Module"
-                val compileModuleTaskName = compileKotlinTask.name + "Module"
 
                 kotlin.sourceSets.create(sourceSetName) {
                     val sourceFile = this.kotlin.find { it.name == "module-info.java" }
-                    val targetFile = compileKotlinTask.destinationDirectory.file("../module-info.class").get().asFile
+                    val targetDirectory = compileKotlinTask.destinationDirectory.map {
+                        it.dir("../${it.asFile.name}Module")
+                    }
 
                     // only configure the compilation if necessary
                     if (sourceFile != null) {
-                        // the default source set depends on this new source set
-                        defaultSourceSet.dependsOn(this)
+                        // register and wire a task to verify module-info.java content
+                        val verifyModuleTask = registerVerifyModuleTask(
```

Comment:
Am I right that this actually will cause the full compilation to happen twice (you've mentioned it somewhere)?
I believe it should be mentioned in the comments then

Reply:
Yes, when running with `check` this is correct.
During pure building it should not happen.
I added that info to the comment now.